src/extension.ts

```diff
@@ -121,10 +121,20 @@ export async function activate(context: ExtensionContext): Promise<void> {
     },
     async () => {
       commands.executeCommand("setContext", "metals:enabled", true);
+
+      const javaHomeConfig = ((jhPath) =>
+        jhPath?.trim() && !path.isAbsolute(jhPath)
+          ? path.resolve(
+              ...[workspace.workspaceFolders?.[0]?.uri.fsPath, jhPath].filter(
+                (s): s is string => !!s
+              )
+            )
+          : jhPath)(
+        workspace.getConfiguration("metals").get<string>("javaHome")
+      );
```

Comment:
I was thinking about little bit more verbose
```typescript
export function getJavaHomeFromConfig(): string | undefined {
  const javaHomePath = workspace
    .getConfiguration("metals")
    .get<string>("javaHome");
  if (javaHomePath?.trim() && !path.isAbsolute(javaHomePath)) {
    const pathSegments = [
      workspace.workspaceFolders?.[0]?.uri.fsPath,
      javaHomePath,
    ].filter((s): s is string => s != null);
    return path.resolve(...pathSegments);
  } else {
    return javaHomePath;
  }
}
```
It's not as concise as what you're proposing but I find it easier to read ;)

Also, `extension.ts` is an enormous file, moving this function to e.g. `util.ts` might be a good idea.

Reply:
Good point about moving this into `util.ts` - I wasn't sure whether it would be better to have as little impact in the scope of the code, or to have a more verbose implementation. I'll make those changes.